src/device/device_ledger.cpp

```diff
@@ -1346,6 +1347,75 @@ namespace hw {
         return true;
     }
 
+    void device_ledger::get_transaction_prefix_hash(const cryptonote::transaction_prefix& tx, crypto::hash& h) {
+      AUTO_LOCK_CMD();
+      
+      int pref_length = 0, pref_offset = 0, offset = 0;
+
+      #ifdef DEBUG_HWDEVICE
+      crypto::hash h_x;
+      crypto::hash h_xx;
+      this->controle_device->get_transaction_prefix_hash(tx,h_x);
+      MDEBUG("get_transaction_prefix_hash [[IN]] h_x/1 "<<h_x);
+      #endif
+    
+      std::ostringstream s_x;
+      binary_archive<true> a_x(s_x);
+      ::serialization::serialize(a_x, const_cast<cryptonote::transaction_prefix&>(tx));
+      pref_length = s_x.str().size();
+      char pref[pref_length];
+      memmove(pref, s_x.str().data(), pref_length);
+
+      offset = set_command_header_noopt(INS_PREFIX_HASH,1);
+
+      //version as varint
+      pref_offset = 0;
+      do {
+        this->buffer_send[offset] = pref[pref_offset];
+        offset += 1;
+        pref_offset += 1;
+      } while (pref[pref_offset]&0x80);
```

Comment:
That looks (ie, I did not check carefully) wrong, if it's monero varints. The high bit tells you whether there is another byte to follow. Here, you check the high bit of a byte you won't use. This can also buffer overrun with crafted data (both read and write).

Reply:
I take the whole varint and pass it to device, which will decode it. This code has been tested with device (app not yep pushed) and it works.

In fact I recompute the prefix hash on device and will this once during the signature. So I send the raw serialization prefix hash to the app.  

OK I will check buffer overrun, in a general way, I assert the the passed argument are well-formed.

Reply:
Until proven otherwise, I think it's working by chance. Imagine you've got a buffer with a varint with value 1, which encodes as just "1". Imagine the next data is a set of 0x80. The code will copy the first byte (1), then increase pref_offset to 1, then check that for 0x80. It matches, so the loop will continue. That is wrong. Is the above wrong ?

Reply:
Seems you are right yes! I rewrite my initial `while` loop to a `do while` one, a bit too quickly!